src/autosave.js

```diff
@@ -0,0 +1,176 @@
+		 * @member {Provider}
+		 */
+		this.provider = undefined;
+
+		/**
+		 * @protected
+		 * @type {Function}
+		 */
+		this._throttledSave = throttle( this._save.bind( this ), 500 );
+
+		/**
+		 * @private
+		 * @type {DomEmitterMixin}
+		 */
+		this._domEmitter = Object.create( DomEmitterMixin );
+
+		/**
+		 * @protected
+		 * @type {Number}
+		 */
+		this._lastDocumentVersion = editor.model.document.version;
+	}
+
+	init() {
+		const editor = this.editor;
+		const doc = editor.model.document;
+		const pendingActions = editor.plugins.get( PendingActions );
+
+		this.listenTo( doc, 'change', this._saveIfDocumentChanged.bind( this ) );
+
+		// Flush on the editor's destroy listener with the highest priority to ensure that
+		// `editor.getData()` will be called before plugins are destroyed.
+		this.listenTo( editor, 'destroy', () => this._save(), { priority: 'highest' } );
+
+		// It's not possible to easy test it because karma uses `beforeunload` event
+		// to warn before full page reload and this event cannot be dispatched manually.
+		/* istanbul ignore next */
+		this._domEmitter.listenTo( window, 'beforeunload', ( evtInfo, domEvt ) => {
+			if ( pendingActions.isPending ) {
+				domEvt.returnValue = pendingActions.first.message;
+			}
+		} );
+	}
+
+	destroy() {
+		this._throttledSave.cancel();
+		this._domEmitter.stopListening();
+		super.destroy();
+	}
+
+	/**
+	 * @protected
+	 */
+	_flush() {
+		this._throttledSave.flush();
+	}
+
+	/**
+	 * Filters out changes in document, that don't impact on the data (e.g. selection changes).
+	 * Quickly returns when finds an operation that potentially changes document's data.
+	 *
+	 * @private
+	 */
+	_saveIfDocumentChanged( evt, batch ) {
+		for ( const delta of batch.deltas ) {
+			for ( const operation of delta.operations ) {
+				const name = operation.name;
+
+				if ( !name || !isUserSelectionOperation( operation ) ) {
+					this._throttledSave();
+
+					return;
+				}
+			}
+		}
+	}
+
+	/**
+	 * If the provider is set and new document version exists,
+	 * `_save()` method creates a pending action and calls `provider.save()` method.
+	 * It waits for the result and then removes the created pending action.
+	 *
+	 * @private
+	 */
+	_save() {
+		if ( !this.provider ) {
+			return;
+		}
+
+		const version = this.editor.model.document.version;
+
+		if ( version <= this._lastDocumentVersion ) {
+			return;
+		}
+
+		this._lastDocumentVersion = version;
+
+		const pendingActions = this.editor.plugins.get( PendingActions );
+		const action = pendingActions.add( 'Saving in progress.' );
+
+		Promise.resolve( this.provider.save() )
+			.then( () => {
+				pendingActions.remove( action );
+			} );
+	}
+}
+
+/**
+ * @typedef Provider
+ */
+
+/**
+ * @function
+ * @name Provider#save
+ * @type {Function}
+ */
+
+/**
+ * @private
+ */
+function isUserSelectionOperation( operation ) {
+	return (
+		operation.name.startsWith( 'user:position' ) ||
+		operation.name.startsWith( 'user:range' )
+	);
+}
```

Comment:
We agreed that in the first implementation we will try to no check anything. This is also, what I put here: https://github.com/ckeditor/ckeditor5-autosave/issues/1#issuecomment-390213527

> First, we need to check if it will be a problem in real cases. Editor blur will not cause selection change, so the user should be able to finish editing if no one will not move selection for 500 ms. It might be good enough.

Reply:
However, more I think about it more I think that it will be an issue and we need a more generic solution. Autosave won't be the only way people want to get the information if data should be saved and we should make it easier. This is why I propose to close https://github.com/ckeditor/ckeditor5-engine/issues/1418 first.